[feature-request] Include name of documented object in `D417` (`undocumented-param`) error message

Just ran `D417` (`undocumented-param`) on the codebase. We haven't turned this on yet because we have so many violations. The errors printed look like this:

```
python_modules/dagster/dagster/_core/definitions/job_definition.py:587:9: D417 Missing argument descriptions in the docstring: `asset_selection`, `run_config`, `run_id`, `tags`
```

It's useful but it would be better if it included the names of the objects that the failing docstrings are attached to:

```
python_modules/dagster/dagster/_core/definitions/job_definition.py:587:9: D417 Missing argument descriptions in the docstring for `JobDefinition`: `asset_selection`, `run_config`, `run_id`, `tags`
```

--- a/crates/ruff/src/rules/pydocstyle/rules/sections.rs
+++ b/crates/ruff/src/rules/pydocstyle/rules/sections.rs
@@ -1187,19 +1187,22 @@ impl AlwaysAutofixableViolation for SectionNameEndsInColon {
 /// - [Google Python Style Guide - Docstrings](https://google.github.io/styleguide/pyguide.html#38-comments-and-docstrings)
 #[violation]
 pub struct UndocumentedParam {
-    pub names: Vec<String>,
+    /// The name of the function being documented.
+    definition: String,
+    /// The names of the undocumented parameters.
+    names: Vec<String>,
 }
 
 impl Violation for UndocumentedParam {
     #[derive_message_formats]
     fn message(&self) -> String {
-        let UndocumentedParam { names } = self;
+        let UndocumentedParam { definition, names } = self;
         if names.len() == 1 {
             let name = &names[0];
-            format!("Missing argument description in the docstring: `{name}`")
+            format!("Missing argument description in the docstring for `{definition}`: `{name}`")
         } else {
             let names = names.iter().map(|name| format!("`{name}`")).join(", ");
-            format!("Missing argument descriptions in the docstring: {names}")
+            format!("Missing argument descriptions in the docstring for `{definition}`: {names}")
         }
     }
 }
@@ -1779,11 +1782,16 @@ fn missing_args(checker: &mut Checker, docstring: &Docstring, docstrings_args: &
     }
 
     if !missing_arg_names.is_empty() {
-        let names = missing_arg_names.into_iter().sorted().collect();
-        checker.diagnostics.push(Diagnostic::new(
-            UndocumentedParam { names },
-            stmt.identifier(),
-        ));
+        if let Some(definition) = docstring.definition.name() {
+            let names = missing_arg_names.into_iter().sorted().collect();
+            checker.diagnostics.push(Diagnostic::new(
+                UndocumentedParam {
+                    definition: definition.to_string(),
+                    names,
+                },
+                stmt.identifier(),
+            ));
+        }
     }
 }
 

--- a/crates/ruff_python_semantic/src/definition.rs
+++ b/crates/ruff_python_semantic/src/definition.rs
@@ -47,6 +47,14 @@ impl<'a> Module<'a> {
             None
         }
     }
+
+    /// Return the name of the module.
+    pub fn name(&self) -> Option<&'a str> {
+        match self.source {
+            ModuleSource::Path(path) => path.last().map(Deref::deref),
+            ModuleSource::File(file) => file.file_stem().and_then(std::ffi::OsStr::to_str),
+        }
+    }
 }
 
 #[derive(Debug, Copy, Clone)]
@@ -72,12 +80,13 @@ pub struct Member<'a> {
 }
 
 impl<'a> Member<'a> {
-    fn name(&self) -> &'a str {
+    /// Return the name of the member.
+    pub fn name(&self) -> Option<&'a str> {
         match &self.stmt {
-            Stmt::FunctionDef(ast::StmtFunctionDef { name, .. }) => name,
-            Stmt::AsyncFunctionDef(ast::StmtAsyncFunctionDef { name, .. }) => name,
-            Stmt::ClassDef(ast::StmtClassDef { name, .. }) => name,
-            _ => unreachable!("Unexpected member kind: {:?}", self.kind),
+            Stmt::FunctionDef(ast::StmtFunctionDef { name, .. })
+            | Stmt::AsyncFunctionDef(ast::StmtAsyncFunctionDef { name, .. })
+            | Stmt::ClassDef(ast::StmtClassDef { name, .. }) => Some(name),
+            _ => None,
         }
     }
 }
@@ -99,6 +108,13 @@ impl Definition<'_> {
                 ..
             })
         )
+    }
+
+    pub fn name(&self) -> Option<&str> {
+        match self {
+            Definition::Module(module) => module.name(),
+            Definition::Member(member) => member.name(),
+        }
     }
 }
 
@@ -134,8 +150,9 @@ impl<'a> Definitions<'a> {
                         MemberKind::Class => {
                             let parent = &definitions[member.parent];
                             if parent.visibility.is_private()
-                                || exports
-                                    .map_or(false, |exports| !exports.contains(&member.name()))
+                                || exports.map_or(false, |exports| {
+                                    member.name().map_or(false, |name| !exports.contains(&name))
+                                })
                             {
                                 Visibility::Private
                             } else {
@@ -163,8 +180,9 @@ impl<'a> Definitions<'a> {
                         MemberKind::Function => {
                             let parent = &definitions[member.parent];
                             if parent.visibility.is_private()
-                                || exports
-                                    .map_or(false, |exports| !exports.contains(&member.name()))
+                                || exports.map_or(false, |exports| {
+                                    member.name().map_or(false, |name| !exports.contains(&name))
+                                })
                             {
                                 Visibility::Private
                             } else {

